### PR TITLE
Refactor Dusk CLI

### PR DESCRIPTION
**Pull request by Stagno**
Sparked by the need to pass arguments to `dawn-opt` / `dawn-codegen` (such as the new ones introduced [here](https://github.com/MeteoSwiss-APN/dawn/pull/1024)).
Enables choosing between outputting SIR (default) / generated code to either stdout (default) / file. Also adds `-verbose` option (when generating code) to make dawn verbose.
`-dump-sir` is removed because it's substituted by just specifying the `-o` option (and not `-generate-code`).

### CODE REVIEW

**Comment by BenWeber42** [comment]
I put more thought into this and I'm not sure if this works out well. Maybe it would be easier to have two cli tools. The simple `dusk` cli tool which calls `dawn4py.compile` and a new `dusk-front` which will always output SIR to `stdout` so that it can be used in a pipe way `dusk-front | dawn-opt | dawn-codegen`.

There's two use cases that I'd like to cover and I think they're somewhat mutually exclusive:

* Use dusk like a compiler: Will generate `out.cpp` unless a different path is specified (similar how gcc generates `a.out` by default), but can also generate `<file-name>.sir` if requested (`--dump-sir`).
* Use dusk in the pipe style: Only SIR should be put on `stdout` (cpp doesn't really make sense). Other output (`--verbose`) should go to `stderr` instead.

The first option would also be useful if we have a few folders of dusk stencils, then we can generate `.cpp` and `.sir` files for all of them easily. Dusk can have multiple output streams for it's various formats (`.sir`, `.cpp`) and logging information (`stdout` & `stderr`).

The second option is definitely good so that one can specify the exact options for `dawn-opt` and `dawn-codegen`. However, then dusk will have only one output stream (`stdout`) for the SIR (`cpp` is questionable on `stdout`) and only one logging information stream (`stderr`).

Basically, `-generate-code` isn't really useful without `-o <out-file>`. These two options are quite different and I think it's better to have different cli tools for them.

**Comment by Stagno** [comment]
> I put more thought into this and I'm not sure if this works out well. Maybe it would be easier to have two cli tools. The simple `dusk` cli tool which calls `dawn4py.compile` and a new `dusk-front` which will always output SIR to `stdout` so that it can be used in a pipe way `dusk-front | dawn-opt | dawn-codegen`.
> 
> There's two use cases that I'd like to cover and I think they're somewhat mutually exclusive:
> 
>     * Use dusk like a compiler: Will generate `out.cpp` unless a different path is specified (similar how gcc generates `a.out` by default), but can also generate `<file-name>.sir` if requested (`--dump-sir`).
> 
>     * Use dusk in the pipe style: Only SIR should be put on `stdout` (cpp doesn't really make sense). Other output (`--verbose`) should go to `stderr` instead.
> 
> 
> The first option would also be useful if we have a few folders of dusk stencils, then we can generate `.cpp` and `.sir` files for all of them easily. Dusk can have multiple output streams for it's various formats (`.sir`, `.cpp`) and logging information (`stdout` & `stderr`).
> 
> The second option is definitely good so that one can specify the exact options for `dawn-opt` and `dawn-codegen`. However, then dusk will have only one output stream (`stdout`) for the SIR (`cpp` is questionable on `stdout`) and only one logging information stream (`stderr`).
> 
> Basically, `-generate-code` isn't really useful without `-o <out-file>`. These two options are quite different and I think it's better to have different cli tools for them. This way we can still have the simple tool for the user, while the other tool covers the more advanced use cases.

That's a good idea. Since we are calling dawn through dawn4py, how can we redirect stdout to stderr for the `dusk-front` tool?

**Comment by BenWeber42** [comment]
Yea good question. In python we could probably hack something like:

```python
import sys

actual_stdout = sys.stdout
sys.stdout = sys.stderr

print("lorem") # should now go to stderr
actual_stdout.write("ipsum") # should go to stdout
```

I'd be surprised if this works when C++ code gets called through pybind. However, we only ever want to put SIR to `stdout`. And I'm not sure if C++ code gets called for building SIR and converting to JSON. Otherwise, there might be some other hacks that could help us :sweat_smile:

**Comment by Stagno** [comment]
SIR is returned as a string (in python) if i'm not mistaken. However dawn will output diagnostics to stdout.

**Comment by BenWeber42** [comment]
Yes, the challenge will be to deal with dawn's output to stdout. However, we have to solve this either way if we want dusk to be usable in a _pipe-style_. What does dawn do when it's used in a pipe-style? `dawn-opt` and `dawn-codegen` would face the same issue right? If they put diagnostics to stdout, it breaks the piped chain.

Worst case scenario, we have to use some wrapper script and hack around with stdout & stderr from a parent process.

**Comment by Stagno** [comment]
Behavior of `dawn-opt` and `dawn-codegen` can be guessed from these 3 lines:
```
Logger info(makeMessageFormatter("INFO"), makeDiagnosticFormatter("INFO"), std::cout, false);
Logger warn(makeMessageFormatter("WARNING"), makeDiagnosticFormatter("WARNING"), std::cout, true);
Logger error(makeMessageFormatter("ERROR"), makeDiagnosticFormatter("ERROR"), std::cerr, true);
```
where `INFO` is outputted only when `-v` is provided:
```
  -v, --verbose          Set verbosity level to info. If set, use -o or --out
                         to redirect IIR.
```
so it shouldn't be a problem (as `-v` would be used only for debugging purposes).
`ERROR` is also ok, as it goes to stderr.
However, `WARNING` goes to stdout along with the main output (IIR/gen. code), therefore "poisoning" it when piping.
Maybe we should consider putting `WARNING` in stderr, it seems the cleanest solution.

**Review by BenWeber42 — APPROVED** [review]
